tailfs/remote.go

```diff
@@ -0,0 +1,390 @@
+
+// Share represents a folder that's shared with remote Tailfs nodes.
+type Share struct {
+	// Name is how this share appears on remote nodes.
+	Name string `json:"name"`
+	// Path is the path to the directory on this machine that's being shared.
+	Path string `json:"path"`
+	// As is the UNIX or Windows username of the local account used for this
+	// share. File read/write permissions are enforced based on this username.
+	As string `json:"who"`
+	// Readers is a list of Tailscale principals that are allowed to read this
+	// share.
+	Readers []string `json:"readers,omitempty"`
+	// Writers is a list of Tailscale principals that are allowed to write to
+	// this share.
+	Writers []string `json:"writers,omitempty"`
+}
+
+// ForRemote is the TailFS filesystem exposed to remote nodes. It provides a
+// unified WebDAV interface to local directories that have been shared.
+type ForRemote interface {
+	// SetFileServerAddr sets the address of the file server to which we
+	// should proxy. This is used on platforms like Windows and MacOS
+	// sandboxed where we can't spawn user-specific sub-processes and instead
+	// rely on the UI application that's already running as an unprivileged
+	// user to access the filesystem for us.
+	SetFileServerAddr(addr string)
+
+	// SetShares sets the complete set of shares exposed by this node. If
+	// AllowShareAs() is true, we will use one subprocess per user to access
+	// the filesystem (see userServer). Otherwise, we will use the file server
+	// configured via SetFileServerAddr.
+	SetShares(shares map[string]*Share)
+
+	// ServeHTTP behaves like the similar method from http.Handler but also
+	// accepts a Permissions map that captures the permissions of the connecting
+	// node.
+	ServeHTTP(permissions Permissions, w http.ResponseWriter, r *http.Request)
+
+	// Close() stops serving the WebDAV content
+	Close() error
+}
+
+func NewFileSystemForRemote(logf logger.Logf) ForRemote {
+	fs := &fileSystemForRemote{
+		logf:        logf,
+		lockSystem:  webdav.NewMemLS(),
+		fileSystems: make(map[string]webdav.FileSystem),
+		userServers: make(map[string]*userServer),
+	}
+	return fs
+}
+
+type fileSystemForRemote struct {
+	logf           logger.Logf
+	lockSystem     webdav.LockSystem
+	fileServerAddr string
+	shares         map[string]*Share
+	fileSystems    map[string]webdav.FileSystem
+	userServers    map[string]*userServer
+	mx             sync.RWMutex
+}
+
+func (s *fileSystemForRemote) SetFileServerAddr(addr string) {
+	s.mx.Lock()
+	s.fileServerAddr = addr
+	s.mx.Unlock()
+}
+
+func (s *fileSystemForRemote) SetShares(shares map[string]*Share) {
+	userServers := make(map[string]*userServer)
+	if AllowShareAs() {
+		// set up per-user server
+		for _, share := range shares {
+			p, found := userServers[share.As]
+			if !found {
+				p = &userServer{
+					logf: s.logf,
+				}
+				userServers[share.As] = p
+			}
+			p.shares = append(p.shares, share)
+		}
+		for _, p := range userServers {
+			go p.runLoop()
```

Comment:
It looks like we are starting these unbounded goroutines each time we create a new share, and I cannot see how they will ever be stopped. That implies if we replace the existing shares on a filesystem, any existing shares will continue merrily running in the background for the rest of the lifetime of the process. 

Maybe there's some reason that isn't a problem? But at least I think we should document that. But longer-term I think we should make a way for these goroutines to be stopped and collected (e.g., plumb in a context maybe, possibly gang them to a waitgroup so we can tell when they're done, etc.) I feel like SetShares should be resopnsible for cleaning up the antecedents.

A separate but parallel question, is there any concern for what happens if someone shares the same thing multiple times, replacing the previous version but potentially competing with it for access? Do we have to worry about the old one stealing work requests from the new one, for example?